Associated Predicates

ranjitjhala opened the pull request:
Here's a starting point, I still need to implement a bunch of checks, namely: checking the sort of the impl-pred against the trait spec, but the below works now, which should be good to make progress on `Vec`. 

In fact, I am tempted to merge this and move onto the Vec stuff _first_, because in this "fast-and-loose" version, we can make some progress _without_ associating the predicates with the `trait` but directly with the `impl`.

```rust
// Step 1 : declare -------------------------------
#[flux::generics(Self as base)]
#[flux::predicate{ f : (Self) -> bool }]
pub trait MyTrait {
    fn method(&self) -> i32;
}

// Step 2 : implement -----------------------------
#[flux::predicate{ f : |x:int| { 0 < x } }] // TODO: check against trait-def
impl MyTrait for i32 {
    fn method(&self) -> i32 {
        10
    }
}

// Step 3 : abstract ------------------------------
#[flux::trusted] // TODO: subtyping with alias_pred on lhs
#[flux::sig(fn<T as base>(&{T[@x] | <T as MyTrait>::f(x)}))] // TODO: check against trait-spec
pub fn bob<T: MyTrait>(x: &T) {
    x.method();
}

// Step 4 : concretize ----------------------------
pub fn test() {
    let z0 = 0;
    bob(&z0); //~ ERROR refinement type
    let z1 = 1;
    bob(&z1);
}
```

nilehmann commented on crates/flux-middle/src/rty/projections.rs:
```diff
@@ -182,10 +218,14 @@ impl FallibleTypeFolder for Normalizer<'_, '_, '_> {
+            TyKind::Alias(AliasKind::Projection, alias_ty) => {
+                self.normalize_projection_ty(alias_ty)
+            }
+            TyKind::Constr(Pred::Alias(alias_pred), ty) => {
+                self.normalize_alias_pred(alias_pred, ty)
+            }
```
Move this into `try_fold_pred` like so

```rust
fn try_fold_pred(&mut self, ty: &Ty) -> Result<Pred, Self::Eror> {
    if let Pred::Alias(pred) = alias_pred {
        self.normalize_alias_pred(alias_pred)
    } else {
        alias_pred.super_try_fold_with(self)
    }
}

nilehmann commented on crates/flux-middle/src/rty/projections.rs:
```diff
@@ -44,6 +44,18 @@ impl<'sess, 'tcx, 'cx> Normalizer<'sess, 'tcx, 'cx> {
         Ok(Normalizer { genv, selcx, def_id: callsite_def_id, param_env })
     }
 
+    fn normalize_alias_pred(&mut self, alias_pred: &AliasPred, ty: &Ty) -> QueryResult<Ty> {
+        if let Some(impl_id) = self.impl_id_of_alias_ty(alias_pred)?
+            && let Some(pred) = self.genv.assoc_predicate_of(impl_id, alias_pred.name)?
+            && let AssocPredicateKind::Impl(body) = pred.kind
+        {
+            let expr = body.replace_bound_exprs(&alias_pred.refine_args);
+            Ok(Ty::constr_expr(expr, ty.clone()))
+        } else {
+            bug!("failed to normalize_alias_pred `{alias_pred:?}`")
+        }
+    }
```
```suggestion
    fn normalize_alias_pred(&mut self, alias_pred: &AliasPred) -> QueryResult<Pred> {
        if let Some(impl_id) = self.impl_id_of_alias_ty(alias_pred)?
            && let Some(pred) = self.genv.assoc_predicate_of(impl_id, alias_pred.name)?
            && let AssocPredicateKind::Impl(body) = pred.kind
        {
            let expr = body.replace_bound_exprs(&alias_pred.refine_args);
            Ok(Pred::Expr(expr))
        } else {
            bug!("failed to normalize_alias_pred `{alias_pred:?}`")
        }
    }
```

nilehmann commented on crates/flux-middle/src/rty/mod.rs:
```diff
@@ -203,6 +203,23 @@ pub struct GeneratorObligPredicate {
 
+#[derive(Debug, Default, Clone)]
+pub struct AssocPredicates {
+    pub predicates: Vec<AssocPredicate>,
```
You should make this cheaply clonable if returned from a query

```suggestion
    pub predicates: List<AssocPredicate>,
```

nilehmann commented on crates/flux-middle/src/rty/projections.rs:
```diff
@@ -44,6 +44,18 @@ impl<'sess, 'tcx, 'cx> Normalizer<'sess, 'tcx, 'cx> {
+            && let Some(pred) = self.genv.assoc_predicate_of(impl_id, alias_pred.name)?
+            && let AssocPredicateKind::Impl(body) = pred.kind
+        {
+            let expr = body.replace_bound_exprs(&alias_pred.refine_args);
```
If this body is represented with a `Binder` the more correct thing to do is to represent `AliasPred` separate from `RefineArgs`. Note also that we are cheating here because `body` is implicitly under an `EarlyBinder` and we are skipping it, but that's ok because generics will only appear in sorts which are being removed because the binder is being removed.

If this comment doesn't make sense. Feel free to ignore, I'll fix it latter.

ranjitjhala replied:
Hmm I don't think I followed, but I did split the `RefineArgs` out of `AliasPred`, by which I presumed you meant (from the HackMD)

```rust
pub enum AliasPred { 
    trait_id: DefId,
    name: Symbol,
    generic_args: GenericArgs,
}

pub enum Pred {
    Expr(Expr),
    Alias(AliasPred, RefineArgs)
}
```

nilehmann reviewed (approved):
I'm ok with merging this after comments are addressed.

Before moving to vectors perhaps we should try making the following definition work without trusting it:

```rust
#[flux::sig(fn<T as base>(&{T[@x] | <T as MyTrait>::f(x)}))]
pub fn bob<T: MyTrait>(x: &T) {
    x.method();
}
```
We just need to encode the associated predicate as an uninterpreted function. We could generalize `refine_tree::NodeKind::Guard` to contain a `Pred` and then do something similar to `FixpointCtxt::collect_sorts` to collect the abstract refinements needed to encode the constraint.

ranjitjhala commented on crates/flux-desugar/src/desugar.rs:
```diff
@@ -177,17 +177,24 @@ enum QPathRes<'a> {
     NumConst(i128),
 }
 
+fn self_res_did(genv: &GlobalEnv, def_id: DefId) -> SelfRes {
```
done.

ranjitjhala commented on crates/flux-desugar/src/desugar.rs:
```diff
@@ -859,6 +906,38 @@ impl<'a, 'tcx> RustItemCtxt<'a, 'tcx> {
+
+            Ok(fhir::AliasPred { trait_id, name: alias_pred.name.name, generic_args, refine_args })
+        } else {
+            Err(self.emit_err(errors::UnresolvedVar::from_path(&alias_pred.trait_id)))
```
Ok, have tweaked it so it says:

```rust
error[FLUX]: cannot find trait `i32` in this scope
 --> crates/flux-tests/tests/neg/error_messages/desugar/assoc_pred00.rs:1:43
  |
1 | #[flux::sig(fn<T as base>(&{T[@x] | <T as i32>::f(x)}, y: i32))] //~ ERROR cannot find trait `i32` in this scope
  |                                           ^^^ not found in this scope
-Ztrack-diagnostics: created at crates/flux-desugar/src/desugar.rs:922:22
```

ranjitjhala replied:
(or did you have something else in mind?)

nilehmann replied:
@ranjitjhala I think the error has nothing to do with not finding `i32`, the problem is that `i32` resolves to something that is not a trait.

ranjitjhala replied:
Ah got it, let me add an explicit test to check the thing is a trait.

ranjitjhala replied:
oh wait, the updated test in fact checks the resolved thing is a trait...

ranjitjhala commented on crates/flux-middle/src/rty/fold.rs:
```diff
@@ -49,6 +50,14 @@ pub trait TypeVisitor: Sized {
         ty.super_visit_with(self)
     }
 
+    fn visit_pred(&mut self, pred: &Pred) -> ControlFlow<Self::BreakTy> {
+        pred.super_visit_with(self)
+    }
+
+    fn visit_alias_pred(&mut self, alias_pred: &AliasPred) -> ControlFlow<Self::BreakTy> {
+        alias_pred.super_visit_with(self)
+    }
+
```
hmm ... looks like removing those breaks a whole bunch of things, so perhaps that is getting used somewhere? In particular, I think its used in the code for `TypeSuperVisitable for Ty` in the `Constr` case which looks (and used to look like)

```rust
            TyKind::Constr(pred, ty) => {
                pred.visit_with(visitor)?;
                ty.visit_with(visitor)
            }
```

except now the `pred.visit_with(...)` is going through the above. [This didn't show up in the diff because the code is unchanged, I'm guessing.]

ranjitjhala commented on crates/flux-middle/src/rty/fold.rs:
```diff
@@ -83,6 +92,14 @@ pub trait FallibleTypeFolder: Sized {
+        pred.try_super_fold_with(self)
+    }
+
+    fn try_fold_alias_pred(&mut self, alias_pred: &AliasPred) -> Result<AliasPred, Self::Error> {
+        alias_pred.try_super_fold_with(self)
+    }
```
(I think we still need them, see previous comment...)

ranjitjhala commented:
Ok I think I addressed all the comments -- except the checking-signatures ("making the following definition work without trusting") -- which I presumed you meant separate from the comments; have factored that into a separate issue and I'll work on it next, thanks!
